### ISSUE TITLE
Phase 2: Create GEMMA Hyperparameter Tuning Script (CODE-ONLY BLUEPRINT)

### ISSUE THREAD

**Issue by SefaGH** [SefaGH/bearish-alpha-bot#373]
### 🎯 Amaç (Goal)

Bu görevin temel amacı, `scripts/tune_regime_models_standalone.py` script'ini bir şablon olarak kullanarak, GEMMA'nın MLP mimarisi için hiperparametre optimizasyonu yapacak **`scripts/tune_gemma_model_standalone.py`** adında yeni bir script **yazmaktır.**

Bu görev, **sadece kod yazımına odaklıdır.** Script'in çalıştırılması veya test edilmesi bu fazın bir parçası DEĞİLDİR. Nihai çıktı, tek ve eksiksiz bir Python script dosyasıdır. Başarı, üretilen kodun bu çizimdeki (blueprint) talimatlara harfiyen uymasıyla ölçülecektir.

### ⚙️ Referans ve Strateji (Reference & Strategy)

*   **Ana Şablon (Blueprint):** Bu görevdeki tüm mantık ve yapı için temel referans dosyası **`scripts/tune_regime_models_standalone.py`**'dir.
*   **Model Mimarisi Referansı:** Oluşturulacak GEMMA modelinin (MLP) katman yapısı için referans alınacak kod, **`scripts/train_all_models.py`** dosyası içindeki `train_gemma_model` fonksiyonunda bulunan `nn.Sequential` bloğudur.

#### **Adım 1: Dosya Yapısını Kopyalama**

1.  `scripts/tune_regime_models_standalone.py` dosyasının içeriğinin tamamını kopyala.
2.  Bu içeriği, **`scripts/tune_gemma_model_standalone.py`** adında yeni bir dosyaya yapıştır. Aşağıdaki tüm değişiklikler bu yeni dosya üzerinde yapılacaktır.

#### **Adım 2: Model Fabrikasını (Model Factory) GEMMA'ya Dönüştürme**

Bu, görevin en kritik adımıdır. `RegimeModelTuner` sınıfı içindeki model oluşturma mantığını değiştireceğiz.

1.  `RegimeModelTuner` sınıfının içindeki `create_lstm_model` metodunu bul.
2.  Metodun adını **`create_gemma_model`** olarak değiştir.
3.  `create_gemma_model` metodunun içindeki `SimpleLSTM` ve `SklearnLSTMWrapper` sınıflarını sil.
4.  Bu metodun içine, aşağıdaki yapıyı birebir kopyalayarak yeni bir `SklearnMLPWrapper` sınıfı tanımla. Bu sınıf, bir MLP (Multi-Layer Perceptron) modeli oluşturacak ve `scikit-learn` uyumlu bir arayüz sağlayacaktır.

    ```python
    # scripts/tune_gemma_model_standalone.py -> create_gemma_model metodu içine eklenecek kod

    import torch
    import torch.nn as nn
    from torch.utils.data import TensorDataset, DataLoader

    class SklearnMLPWrapper:
        """Sklearn-compatible wrapper for a PyTorch MLP model with balanced training."""
        
        def __init__(self, input_size, hidden_size, num_layers, num_classes, dropout, 
                     learning_rate, weight_decay, batch_size, class_weights=None):
            
            # ==================== DEĞİŞEN ANA KISIM BURASI ====================
            # Dinamik MLP (Feed-forward) modelini oluştur
            layers = []
            layers.append(nn.Linear(input_size, hidden_size))
            layers.append(nn.ReLU())
            layers.append(nn.Dropout(dropout))
            
            for _ in range(num_layers - 1):
                layers.append(nn.Linear(hidden_size, hidden_size))
                layers.append(nn.ReLU())
                layers.append(nn.Dropout(dropout))
            
            layers.append(nn.Linear(hidden_size, num_classes))
            self.model = nn.Sequential(*layers)
            # =================================================================

            if class_weights is not None:
                self.criterion = nn.CrossEntropyLoss(weight=torch.FloatTensor(class_weights))
            else:
                self.criterion = nn.CrossEntropyLoss()
            
            self.optimizer = torch.optim.Adam(
                self.model.parameters(),
                lr=learning_rate,
                weight_decay=weight_decay  # L2 Regularization
            )
            self.batch_size = batch_size
            self.num_epochs = 25
            self.patience = 7
            self.min_delta = 0.001

        # ================== BU KISIMLAR DEĞİŞMEYECEK ==================
        # 'fit', 'predict', ve 'score' metodlarını, orijinal
        # 'SklearnLSTMWrapper' sınıfından DEĞİŞTİRMEDEN kopyala.
        # Erken durdurma (early stopping) mantığı dahil olmak üzere
        # tüm eğitim döngüsü aynı kalmalıdır.
        # =============================================================
        def fit(self, X, y):
            """Sklearn-style fit with validation split."""
            # ... (Orijinal koddaki gibi, DEĞİŞTİRME) ...
            return self
        
        def predict(self, X):
            """Sklearn-style predict method."""
            # ... (Orijinal koddaki gibi, DEĞİŞTİRME) ...
            # ÖNEMLİ NOT: LSTM'deki 'x.unsqueeze(1)' satırı burada OLMAMALIDIR.
            # MLP modeli 2D veri bekler, 3D değil.
            self.model.eval()
            with torch.no_grad():
                X_tensor = torch.FloatTensor(X)
                outputs = self.model(X_tensor)
                _, predicted = torch.max(outputs, 1)
                return predicted.numpy()
        
        def score(self, X, y):
            """Sklearn-style score method."""
            # ... (Orijinal koddaki gibi, DEĞİŞTİRME) ...
    
    # Metodun sonunda, bu yeni wrapper'ı return et
    return SklearnMLPWrapper(
        input_size=params.get('input_size', 83), # Beklenen özellik sayısı 83
        hidden_size=params['hidden_size'],
        num_layers=params['num_layers'],
        num_classes=params.get('num_classes', 3),
        dropout=params['dropout'],
        learning_rate=params.get('learning_rate', 0.001),
        weight_decay=params.get('weight_decay', 0.0), # MLP için weight_decay 0 olabilir
        batch_size=params.get('batch_size', 64),
        class_weights=params.get('class_weights', None)
    )
    ```

5.  `tune_model` metodu içindeki `model_factory`'yi, yeni **`create_gemma_model`**'i çağıracak şekilde güncelle.
    ```python
    # tune_model metodu içindeki bu bloğu bul ve güncelle
    def model_factory(params):
        if model_type == 'gemma':  # <-- 'lstm' yerine 'gemma'
            params['input_size'] = X_clean.shape[1]
            params['num_classes'] = num_classes
            params['class_weights'] = class_weights
            return self.create_gemma_model(params) # <-- Fonksiyon adını değiştir
        else:
            raise ValueError(f"Unknown model: {model_type}")
    ```

#### **Adım 3: Hiperparametre ve Argümanları Ayarlama**

1.  `OptunaModelTuner` sınıfı bu script'te değil, `scripts/utils/optuna_tuner.py`'de tanımlanmış olabilir. Eğer öyleyse, bu adımı atla. Eğer bu script'in içinde bir "arama uzayı" tanımı varsa, onu aşağıdaki gibi MLP'ye uygun hale getir:
    *   `hidden_size`: `trial.suggest_int('hidden_size', 32, 128, log=True)`
    *   `num_layers`: `trial.suggest_int('num_layers', 2, 4)`
    *   `dropout`: `trial.suggest_uniform('dropout', 0.2, 0.6)`
    *   `learning_rate`: `trial.suggest_loguniform('learning_rate', 1e-4, 1e-2)`
    *   `weight_decay`: `trial.suggest_loguniform('weight_decay', 1e-6, 1e-3)`
    *   `batch_size`: `trial.suggest_categorical('batch_size', [32, 64, 128])`

2.  Script'in en altındaki `main()` fonksiyonunda, `argparse` bölümünü güncelle. `--model` argümanının artık sadece `'gemma'` olmasını sağla.
    ```python
    # main() fonksiyonu içindeki bu bloğu bul ve güncelle
    parser = argparse.ArgumentParser()
    parser.add_argument('--model', default='gemma', choices=['gemma']) # Sadece gemma
    # ... (diğer argümanlar aynı kalabilir)
    ```

3.  `_save_results` metodunda, kaydedilen `.json` dosyasının adının `gemma_tuning_` ile başladığından emin ol.
    ```python
    # _save_results metodu içindeki bu satırı bul ve güncelle
    filename = f"gemma_tuning_{datetime.utcnow().strftime('%Y%m%d_%H%M%S')}.json"
    ```

### ✅ Başarı Kriterleri (Code-Level Success Criteria)

- [ ] `scripts/tune_gemma_model_standalone.py` adında yeni bir dosya oluşturulmuş olmalıdır.
- [ ] Dosya içinde `create_gemma_model` adında bir metod bulunmalıdır.
- [ ] Bu metodun içinde, bir `nn.Sequential` MLP modeli oluşturan `SklearnMLPWrapper` sınıfı tanımlanmalıdır.
- [ ] `SklearnMLPWrapper` içindeki `predict` metodu, LSTM'e özel olan `x.unsqueeze(1)` satırını **içermemelidir.**
- [ ] `tune_model` metodu, `model_type` olarak `'gemma'`'yı kontrol etmeli ve `create_gemma_model`'i çağırmalıdır.
- [ ] `main()` fonksiyonu, komut satırı argümanı olarak `'gemma'`'yı varsayılan olarak ayarlamalıdır.
- [ ] `_save_results` metodu, çıktı dosyasını `gemma_tuning_...` olarak adlandırmalıdır.
- [ ] Kod, sentaktik olarak hatasız olmalı ve herhangi bir çalıştırma veya test gerektirmemelidir.
